### PR TITLE
Add suffix to the job name if it already exists

### PR DESCRIPTION
**Pull request by omer-dayan**

### CODE REVIEW

File: cmd/job/submit/submit_mpi.go [hunk 1]
```diff
@@ -148,10 +148,11 @@ func submitMPIJob(cmd *cobra.Command, args []string, submitArgs *submitMPIJobArg
 	// the master is also considered as a worker
 	// submitArgs.WorkerCount = submitArgs.WorkerCount - 1
 
-	err = workflow.SubmitJob(submitArgs.Name, submitArgs.Mode, submitArgs.Namespace, submitArgs, *configValues, mpijob_chart, client.GetClientset(), dryRun)
+	mutatedName, err := workflow.SubmitJob(submitArgs.Name, submitArgs.Mode, submitArgs.Namespace, submitArgs, *configValues, mpijob_chart, client.GetClientset(), dryRun)
 	if err != nil {
 		return err
 	}
+	submitArgs.Name = mutatedName
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit_mpi.go:155]: Why not change the name inside the method `SubmitJob`?

File: cmd/job/submit/submit_runai.go [hunk 2]
```diff
@@ -245,10 +245,12 @@ func submitRunaiJob(args []string, submitArgs *submitRunaiJobArgs, clientset kub
 	}
 
 	handleRunaiJobCRD(submitArgs, runaiclientset)
-	err := workflow.SubmitJob(submitArgs.Name, trainer.DefaultRunaiTrainingType, submitArgs.Namespace, submitArgs, *configValues, runaiChart, clientset, dryRun)
+	submitArgs.Labels[kubectl.BaseNameLabel] = submitArgs.Name
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit_runai.go:248]: Why only for runaijobs? let's do it for all jobs.
Let's move it to the `SubmitJob` method

File: pkg/workflow/workflow.go [hunk 3]
```diff
@@ -125,7 +128,32 @@ func generateJobFiles(name string, namespace string, values interface{}, environ
...
+
+		if !jobExists {
+			return jobName, name, generatedJobFiles, nil
+		}
+	}
+
+	return "", "", nil, fmt.Errorf("The job %s already exists, please delete it first. use '%s delete %s'", name, config.CLIName, name)
+}
+
+func SubmitJob(name string, trainingType string, namespace string, values interface{}, environmentValues string, chart string, clientset kubernetes.Interface, dryRun bool) (string, error) {
```
Comment (yodarshafrir1) [inline pkg/workflow/workflow.go:157]: Let's move the submitargs to this method and change the name of the workload here

File: pkg/util/kubectl/kubectl.go [hunk 4]
```diff
@@ -262,6 +265,40 @@ func LabelAppConfigmap(jobName, namespace, label string) (err error) {
 	return err
 }
 
+func CountJobsByBaseName(baseName, namespace string) (int, error) {
+	binary, err := exec.LookPath(kubectlCmd[0])
+	if err != nil {
+		return 0, err
+	}
+
+	args := []string{"get", "runaijob.run.ai"}
```
Comment (yodarshafrir1) [inline pkg/util/kubectl/kubectl.go:274]: Let's support all workloads type?

File: pkg/util/kubectl/kubectl.go [hunk 5]
```diff
@@ -262,6 +265,40 @@ func LabelAppConfigmap(jobName, namespace, label string) (err error) {
...
+	args = append(args, "--selector=" + BaseNameLabel + "=" + baseName)
```
Comment (yodarshafrir1) [inline pkg/util/kubectl/kubectl.go:275]: Great job using selectors here!!

File: pkg/util/kubectl/kubectl.go [hunk 6]
```diff
@@ -262,6 +265,40 @@ func LabelAppConfigmap(jobName, namespace, label string) (err error) {
...
+	args = append(args, "-o", "json")
+	args = util.AddNamespaceToArgs(args, namespace)
+
+	log.Debugf("kubectl %v", args)
+
+	cmd := exec.Command(binary, args...)
+	env := os.Environ()
+	if types.KubeConfig != "" {
```
Comment (yodarshafrir1) [inline pkg/util/kubectl/kubectl.go:283]: Why is this needed?

File: pkg/util/kubectl/kubectl.go [hunk 7]
```diff
@@ -262,6 +265,40 @@ func LabelAppConfigmap(jobName, namespace, label string) (err error) {
...
+	}
+
+	args := []string{"get", "runaijob.run.ai"}
+	args = append(args, "--selector=" + BaseNameLabel + "=" + baseName)
+	args = append(args, "-o", "json")
+	args = util.AddNamespaceToArgs(args, namespace)
+
+	log.Debugf("kubectl %v", args)
+
+	cmd := exec.Command(binary, args...)
```
Comment (yodarshafrir1) [inline pkg/util/kubectl/kubectl.go:281]: Why did you decide to use `kubectl` command and not API server calls? 
here is an example:
https://github.com/run-ai/runai-cli/blob/master/cmd/trainer/trainer_runai.go#L44

File: pkg/workflow/workflow.go [hunk 8]
```diff
@@ -125,7 +128,32 @@ func generateJobFiles(name string, namespace string, values interface{}, environ
...
+	count, err := kubectl.CountJobsByBaseName(name, namespace)
+	if err != nil {
+		return "", "", nil, err
+	}
+
+	for i := 0; i < forceSubmitRetries; i++ {
+		jobSuffix := strconv.Itoa(count + i - 1)
+		jobName = jobName + "-" + jobSuffix
+		name = name + "-" + jobSuffix
+		generatedJobFiles, err := generateJobFiles(name, namespace, values, environmentValues, chart)
```
Comment (yodarshafrir1) [inline pkg/workflow/workflow.go:142]: there's some duplication code here with normal job creation, can we extract to methods?

File: pkg/workflow/workflow.go [hunk 9]
```diff
@@ -125,7 +128,32 @@ func generateJobFiles(name string, namespace string, values interface{}, environ
...
+		}
+
+		jobExists, err := kubectl.CheckIfAppInfofileContentsExists(generatedJobFiles.appInfoFileName, namespace)
+
+		if !jobExists {
+			return jobName, name, generatedJobFiles, nil
+		}
+	}
+
+	return "", "", nil, fmt.Errorf("The job %s already exists, please delete it first. use '%s delete %s'", name, config.CLIName, name)
```
Comment (yodarshafrir1) [inline pkg/workflow/workflow.go:154]: I'd rephrase the error here.
we claim to have this feature, maybe more in the directions of: failed to create job please submit again.

File: pkg/workflow/workflow.go [hunk 10]
```diff
@@ -136,19 +164,23 @@ func SubmitJob(name string, trainingType string, namespace string, values interf
...
 			jobExists, err := kubectl.CheckIfAppInfofileContentsExists(jobFiles.appInfoFileName, namespace)
 
 			if err != nil {
-				return err
+				return "", err
 			}
 
 			if jobExists {
-				return fmt.Errorf("The job %s already exists, please delete it first. use '%s delete %s'", name, config.CLIName, name)
+				jobName, name, jobFiles, err = forceGenerateJobFiles(name, jobName, namespace, environmentValues, chart, values)
```
Comment (yodarshafrir1) [inline pkg/workflow/workflow.go]: Not a must: This area is a bit confusing, don't be afraid to make some small refactor here

**Review by yodarshafrir1 — CHANGES_REQUESTED** [review]

File: pkg/util/kubectl/kubectl.go (hunk 6 above)
Comment (omer-dayan) [reply pkg/util/kubectl/kubectl.go]: Removed, really not needed

File: cmd/job/submit/submit_mpi.go (hunk 1 above)
Comment (omer-dayan) [reply cmd/job/submit/submit_mpi.go]: Done

File: cmd/job/submit/submit_runai.go (hunk 2 above)
Comment (omer-dayan) [reply cmd/job/submit/submit_runai.go]: Done

File: pkg/util/kubectl/kubectl.go (hunk 7 above)
Comment (omer-dayan) [reply pkg/util/kubectl/kubectl.go:281]: Couldnt find a way to get the data, because we are using CDRs

File: pkg/util/kubectl/kubectl.go (hunk 4 above)
Comment (omer-dayan) [reply pkg/util/kubectl/kubectl.go:274]: Arent they all of them under runaijob.run.ai?

Comment (yodarshafrir1) [reply pkg/util/kubectl/kubectl.go:274]: No, we have mpi jobs, statefulsets and runaijobs

File: pkg/util/kubectl/kubectl.go (hunk 7 above)
Comment (yodarshafrir1) [reply pkg/util/kubectl/kubectl.go:281]: let's have a zoom session and i will show how can we handle this

File: pkg/workflow/workflow.go [hunk 11]
```diff
@@ -125,8 +128,43 @@ func generateJobFiles(name string, namespace string, values interface{}, environ
 
 }
 
-func SubmitJob(name string, trainingType string, namespace string, values interface{}, environmentValues string, chart string, clientset kubernetes.Interface, dryRun bool) error {
+func forceGenerateJobFiles(name *string , namespace, environmentValues, chart string, values interface{}, countFunc getJobCount) (*JobFiles, error) {
```
Comment (yodarshafrir1) [inline pkg/workflow/workflow.go:131]: Can we pass submitArgs.Labels instead of this interface and assign the labels?

File: pkg/workflow/workflow.go [hunk 12]
```diff
@@ -125,8 +128,43 @@ func generateJobFiles(name string, namespace string, values interface{}, environ
...
+	}
+
+	return nil, fmt.Errorf("Could not submit %s. Pleas try again", name)
+}
+
+type submitValues interface {
+	AddLabel(key, value string)
+}
+
+type getJobCount func(name, namespace string) (int, error)
```
Comment (yodarshafrir1) [inline pkg/workflow/workflow.go:162]: Can you put this on the top of the file please
Please also add it suffix for `Fn`

File: cmd/job/submit/submit_runai.go [hunk 13]
```diff
@@ -244,8 +244,15 @@ func submitRunaiJob(args []string, submitArgs *submitRunaiJobArgs, clientset kub
 		return err2
 	}
 
+	countJobFunc := func(name, namespace string) (int, error) {
+		list, err := runaiclientset.RunV1().RunaiJobs(namespace).List(metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", workflow.BaseNameLabel, name)})
+		if err != nil {
+			return 0, err
+		}
+		return len(list.Items), nil
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit_runai.go:252]: what happens on the following scenario:
submit job
submit job
submit job

delete job
submit job

will it work?

File: cmd/job/submit/submit_runai.go [hunk 14]
```diff
@@ -244,8 +244,15 @@ func submitRunaiJob(args []string, submitArgs *submitRunaiJobArgs, clientset kub
 		return err2
 	}
 
+	countJobFunc := func(name, namespace string) (int, error) {
+		list, err := runaiclientset.RunV1().RunaiJobs(namespace).List(metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", workflow.BaseNameLabel, name)})
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit_runai.go:248]: please support interactive jobs as well
you can submit them by adding `--interactive` flag

File: pkg/workflow/workflow.go [hunk 15]
```diff
@@ -125,8 +128,43 @@ func generateJobFiles(name string, namespace string, values interface{}, environ
...
+
+		jobExists, err := kubectl.CheckIfAppInfofileContentsExists(generatedJobFiles.appInfoFileName, namespace)
+
+		if !jobExists {
+			*name = currentName
+			return generatedJobFiles, nil
+		}
+	}
+
+	return nil, fmt.Errorf("Could not submit %s. Pleas try again", name)
```
Comment (yodarshafrir1) [inline pkg/workflow/workflow.go:155]: `Pleas` -> `Please`

File: cmd/job/submit/submit_mpi.go [hunk 16]
```diff
@@ -147,8 +148,14 @@ func submitMPIJob(cmd *cobra.Command, args []string, submitArgs *submitMPIJobArg
 
 	// the master is also considered as a worker
 	// submitArgs.WorkerCount = submitArgs.WorkerCount - 1
-
-	err = workflow.SubmitJob(submitArgs.Name, submitArgs.Mode, submitArgs.Namespace, submitArgs, *configValues, mpijob_chart, client.GetClientset(), dryRun)
+	countJobFunc := func(name, namespace string) (int, error) {
+		list, err := client.GetClientset().CoreV1().Pods(namespace).List(metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", workflow.BaseNameLabel, name)})
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit_mpi.go:152]: why pods? get MPIJobs no?

File: cmd/job/submit/submit_mpi.go [hunk 17]
```diff
@@ -147,8 +148,14 @@ func submitMPIJob(cmd *cobra.Command, args []string, submitArgs *submitMPIJobArg
 
 	// the master is also considered as a worker
 	// submitArgs.WorkerCount = submitArgs.WorkerCount - 1
-
-	err = workflow.SubmitJob(submitArgs.Name, submitArgs.Mode, submitArgs.Namespace, submitArgs, *configValues, mpijob_chart, client.GetClientset(), dryRun)
+	countJobFunc := func(name, namespace string) (int, error) {
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit_mpi.go:151]: Please move the comment you have here to the relevant place

**Review by yodarshafrir1 — CHANGES_REQUESTED** [review]
Looking good, please see my notes

File: cmd/job/submit/submit_runai.go (hunk 13 above)
Comment (yodarshafrir1) [reply cmd/job/submit/submit_runai.go]: @omer-dayan

File: pkg/workflow/workflow.go [hunk 18]
```diff
@@ -125,8 +130,37 @@ func generateJobFiles(name string, namespace string, values interface{}, environ
...
-func SubmitJob(name string, trainingType string, namespace string, values interface{}, environmentValues string, chart string, clientset kubernetes.Interface, dryRun bool) error {
+func forceGenerateJobFiles(name *string , namespace, environmentValues, chart string, values interface{}, countFunc getJobCountFunc) (*JobFiles, error) {
+	count, err := countFunc(*name, namespace)
+	if err != nil {
+		return nil, err
+	}
+	initialName := *name
+
+	for i := 0; i < forceSubmitRetries; i++ {
+		jobSuffix := strconv.Itoa(count + i - 1)
```
Comment (yodarshafrir1) [inline pkg/workflow/workflow.go:141]: Not sure i understand why the -1

File: pkg/workflow/workflow.go [hunk 19]
```diff
@@ -125,8 +130,37 @@ func generateJobFiles(name string, namespace string, values interface{}, environ
 
 }
 
-func SubmitJob(name string, trainingType string, namespace string, values interface{}, environmentValues string, chart string, clientset kubernetes.Interface, dryRun bool) error {
+func forceGenerateJobFiles(name *string , namespace, environmentValues, chart string, values interface{}, countFunc getJobCountFunc) (*JobFiles, error) {
```
Comment (yodarshafrir1) [inline pkg/workflow/workflow.go:133]: force can miss lead here.
let's rename to `generateJobFilesWithNewName` or something similar

File: pkg/workflow/workflow.go [hunk 20]
```diff
@@ -125,8 +130,37 @@ func generateJobFiles(name string, namespace string, values interface{}, environ
...
+	count, err := countFunc(*name, namespace)
+	if err != nil {
+		return nil, err
+	}
+	initialName := *name
+
+	for i := 0; i < forceSubmitRetries; i++ {
```
Comment (yodarshafrir1) [inline pkg/workflow/workflow.go:140]: rename this as well please

File: pkg/workflow/workflow.go (hunk 18 above)
Comment (yodarshafrir1) [reply pkg/workflow/workflow.go:141]: https://github.com/run-ai/runai-cli/pull/28#discussion_r501758946 I still think this will not work well

**Review by yodarshafrir1 — CHANGES_REQUESTED** [review]
Looking good, almost done

File: cmd/job/submit/submit_mpi.go [hunk 21]
```diff
@@ -129,26 +133,43 @@ func (submitArgs *submitMPIJobArgs) addMPITolerations() {
...
+			occupiedIndexesMap[item.Labels[workflow.JobFamilyIndex]] = true
+		}
+
+		jobCount := len(list.Items)
+		for i:= 1; i <= jobCount; i++ {
+			if !occupiedIndexesMap[strconv.Itoa(i)] {
+				return i, nil
+			}
+		}
+		return jobCount, nil
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit_mpi.go:167]: shouldn't we return `jobCount+1`?

**Review by yodarshafrir1 — CHANGES_REQUESTED** [review]

File: cmd/job/submit/submit_mpi.go [hunk 22]
```diff
@@ -82,6 +84,7 @@ func NewRunaiSubmitMPIJobCommand() *cobra.Command {
 	}
 
 	command.Flags().IntVar(&submitArgs.NumberProcesses, "processes", 1, "Number of distributed training processes.")
+	flags.AddBoolNullableFlag(command.Flags(), &(submitArgs.generateName), "generate-name", "", "Allow the CLI to change the name of the job if the job name already exists")
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit_mpi.go:87]: Maybe this is more suitable: 
`Add suffix to the job name if a job with the submitted name already exists`
@yarongol What do you think?

Comment (yarongol) [reply cmd/job/submit/submit_mpi.go:87]: I don't have any context @yodarshafrir1  on what this pull request is about.
Why do we want to add a suffix to the job if it already exists? It can lead to confusion where you don't read the output and think that the job you created is the original job.
Why not just return an error "Job already exists, please use a different name or delete the original job"
